arith/mersenne31/src/m31_ext6.rs

```diff
@@ -0,0 +1,327 @@
+use arith::{field_common, ExtensionField, FFTField, Field};
+use ethnum::U256;
+use rand::RngCore;
+use serdes::{ExpSerde, SerdeResult};
+use std::{
+    io::{Read, Write},
+    iter::{Product, Sum},
+    ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+};
+
+use crate::{m31::mod_reduce_u32, M31Ext3, M31};
+
+#[derive(Debug, Clone, Copy, Default, Hash, PartialEq, Eq)]
+pub struct M31Ext6 {
+    pub v: [M31Ext3; 2],
+}
+
+field_common!(M31Ext6);
+
+impl ExpSerde for M31Ext6 {
+    const SERIALIZED_SIZE: usize = (32 / 8) * 6;
+
+    #[inline(always)]
+    fn serialize_into<W: Write>(&self, mut writer: W) -> SerdeResult<()> {
+        self.v[0].serialize_into(&mut writer)?;
+        self.v[1].serialize_into(&mut writer)?;
+        Ok(())
+    }
+
+    // FIXME: this deserialization function auto corrects invalid inputs.
+    // We should use separate APIs for this and for the actual deserialization.
+    // NOTE(HS): directly copied from M31ext3 impl
+    #[inline(always)]
+    fn deserialize_from<R: Read>(mut reader: R) -> SerdeResult<Self> {
+        Ok(Self {
+            v: [
+                M31Ext3::deserialize_from(&mut reader)?,
+                M31Ext3::deserialize_from(&mut reader)?,
+            ],
+        })
+    }
+}
+
+impl Field for M31Ext6 {
+    const NAME: &'static str = "Mersenne 31 Extension 6";
+
+    const SIZE: usize = 32 / 8 * 6;
+
+    const FIELD_SIZE: usize = 32 * 6;
+
+    const ZERO: Self = M31Ext6 {
+        v: [M31Ext3::ZERO, M31Ext3::ZERO],
+    };
+
+    const ONE: Self = M31Ext6 {
+        v: [M31Ext3::ONE, M31Ext3::ZERO],
+    };
+
+    const INV_2: M31Ext6 = M31Ext6 {
+        v: [M31Ext3::INV_2, M31Ext3::ZERO],
+    };
+
+    const MODULUS: U256 = M31::MODULUS;
+
+    #[inline(always)]
+    fn zero() -> Self {
+        Self::ZERO
+    }
+
+    #[inline(always)]
+    fn is_zero(&self) -> bool {
+        self.v[0].is_zero() && self.v[1].is_zero()
+    }
+
+    #[inline(always)]
+    fn one() -> Self {
+        Self::ONE
+    }
+
+    fn random_unsafe(mut rng: impl RngCore) -> Self {
+        Self {
+            v: [
+                M31Ext3::random_unsafe(&mut rng),
+                M31Ext3::random_unsafe(&mut rng),
+            ],
+        }
+    }
+
+    fn random_bool(mut rng: impl RngCore) -> Self {
+        M31Ext6 {
+            v: [M31Ext3::random_bool(&mut rng), M31Ext3::ZERO],
+        }
+    }
+
+    fn inv(&self) -> Option<Self> {
```

Comment:
i don't really understand this algorithm...

Reply:
oh, I was doing something tricky, gotta explain here:

The M31Ext3Ext2, i.e., M31Ext6, is $F_{p^6} = F_{p^3} [Y] / (Y^2 +  2)$, and we let $a + b Y \in F_{p^6}$ and $c + d Y \in F_{p^6}$ are mutually inverse each other.

Then $(a + bY) (c + dY) \equiv ac + (bc + ad) Y + bd Y^2 \equiv (ac - 2bd) + (bc + ad) Y \mod Y^2 + 2$.  Now we want $ac - 2bd = 1$ and $bc + ad = 0$.

We can make the second condition true first - suppose we have $a + bY \in F_{p^6}$, $-a + bY$ satisfies, then we normalize back the scalar field back to 1.

$(a + bY) (-a + bY) \equiv (-a^2 - 2b^2) \mod Y^2 + 2$, then we do the `normalize`  variable on line 100.

Eventually, $n = (-a^2 - 2b^2)^{-1} \in F_{p^3}$, then result is $n \times (-a + bY) \in F_{p^6}$.

Reply:
the whole point is to avoid the huge exp from Fermat little theorem...